tree/ntupleutil/v7/src/RNTupleImporter.cxx

```diff
@@ -13,16 +13,333 @@
  * For the list of contributors see $ROOTSYS/README/CREDITS.             *
  *************************************************************************/
 
+#include <ROOT/RError.hxx>
+#include <ROOT/RField.hxx>
+#include <ROOT/RNTuple.hxx>
 #include <ROOT/RNTupleImporter.hxx>
+#include <ROOT/RNTupleOptions.hxx>
 #include <ROOT/RNTupleUtil.hxx>
+#include <ROOT/RPageStorage.hxx>
+#include <ROOT/RPageStorageFile.hxx>
 #include <ROOT/RStringView.hxx>
 
-#include <memory>
+#include <TBranch.h>
+#include <TClass.h>
+#include <TDataType.h>
+#include <TLeaf.h>
+#include <TLeafC.h>
+#include <TLeafElement.h>
+#include <TLeafObject.h>
 
-std::unique_ptr<ROOT::Experimental::RNTupleImporter>
-ROOT::Experimental::RNTupleImporter::Create(std::string_view /* sourceFile */, std::string_view /* treeName */,
-                                            std::string_view /* destFile */)
+#include <cassert>
+#include <cstdint>
+#include <cstring>
+#include <iostream>
+#include <utility>
+
+namespace {
+
+class RDefaultProgressCallback : public ROOT::Experimental::RNTupleImporter::RProgressCallback {
+private:
+   std::uint64_t fNbytesLast = 0;
+
+public:
+   virtual ~RDefaultProgressCallback() {}
+   void Call(std::uint64_t nbytesWritten, std::uint64_t neventsWritten) final
+   {
+      // Report if more than 50MB (compressed) where written since the last status update
+      if (nbytesWritten < (fNbytesLast + 50 * 1000 * 1000))
+         return;
+      std::cout << "Wrote " << nbytesWritten / 1000 / 1000 << "MB, " << neventsWritten << " entries" << std::endl;
+      fNbytesLast = nbytesWritten;
+   }
+
+   void Finish(std::uint64_t nbytesWritten, std::uint64_t neventsWritten) final
+   {
+      std::cout << "Done, wrote " << nbytesWritten / 1000 / 1000 << "MB, " << neventsWritten << " entries" << std::endl;
+   }
+};
+
+} // anonymous namespace
+
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::RNTupleImporter::RCStringTransformation::Transform(std::int64_t /* entry */,
+                                                                       const RImportBranch &branch, RImportField &field)
+{
+   *reinterpret_cast<std::string *>(field.fFieldBuffer) = reinterpret_cast<const char *>(branch.fBranchBuffer.get());
+   return RResult<void>::Success();
+}
+
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::RNTupleImporter::RLeafArrayTransformation::Transform(std::int64_t entry,
+                                                                         const RImportBranch &branch,
+                                                                         RImportField &field)
+{
+   if (entry != fEntry) {
+      fEntry = entry;
+      fNum = 0;
+   }
```

Comment:
Did you consider using an explicit `Reset` at the end of the entry rather than this pattern?

Reply:
That's much nicer! [Done](https://github.com/root-project/root/pull/11825/commits/3d11c894afd80067afe07e596a14763ebad4d69e)